### ISSUE TITLE
Research on topic auto grading

### ISSUE THREAD

**Issue by Andre-Forberg** [GatorEducator/gatorgrader#211]
Hi everyone,

my name is André Forberg, and I am a master student at Westsächsische Hochschule Zwickau (western saxony university) in Germany.
We are currently working on a project with the topic 'auto grading' and our task is to evaluate gatorgrader.
We would like to ask you some questions regarding the problems we have encountered with installation and testing. We contacted @gkapfham via email, and he asked to raise these discussion points on the GatorGrader issue tracker.

### Our Project

The next part is an introduction of ourselfes. You may skip this, it contains a description of our project and the main goal, along with an example.

We are:
a team of students who seek to get the degree Master of Science, and in this semester we have a project with the topic 'auto grading'. We are currently 14 people, which are divided into subteams with the purpose of researching every current auto grading system (web based, static code analysis, building tools, and so on).
Our subteam 'gatorgrader' consists of 3 people (Nagima, Begaiym and me).

The topic:
The project idea came from our CS teacher, who has:
 - several courses per semester (basics of programming I, basics of programming II and development of distributed systems; all of these courses are based on JAVA)
 - several programming tasks per course (usually 2-3 tasks)
 - several projects per course (approximately 30 students, 2 students per project (working together as a group))
which he has to grade.
So this will be 3 times 3 times 15 projects per semester (135!).

The problem:
 - each projects is commited to a git server. Our teacher has to checkout each project, which doesn't mean it is working on the first try. It mostly has to be adjust in order to work (e.g. missing libaries, wrong java-version, ...). These adjustments are very time consuming.
 - When the project is working, the analysis of source code and functionality begins (is the setup of the project as intended, is a logging mechanism included, are the unit tests runnable, ...). This is also very time consuming.
 
The goal: 
 - We want to automate as many of these problems as possible.
 - The checkout and build part can be done by a CI tool, e.g. Jenkins.
 - The analysis of the source code is a bit more tricky, but gatorgrader looks promising for the fulfillment of this goal. 

Example of a project:
To clarify that this will be more than just a working 'hello world!' project, here is a typical scenario:
The student is required to create a pizza order service, via spring boot.
It is not intended as a productive system, but the use cases have to work:
 - Login website
 - user management website (with roles 'customer' and 'admininstrator')
 - list of pizzas website (pizzas are JAVA objects, created from entries in a database)
 - shopping cart website (one cart object for the http session)
 - overview website (summarizes the prices of all pizzas)
 - payment website (payment of the order via a service like paypal (only simulated))
 - order completion website (overview over pizzas, price, payment system & randomized arrival time ;) )

In this scenario, gatorgrader should check:
 - is the entity-boundaries-service programming paradigma ensured (has the project specific packages with these terms)
 - is every method which is public and annotated with some of the common spring annotations equipped with a logging call
 - are the repositories (automatic generated) implementations of JpaRepository
 - is there at least 1 call in the repository which is interpretable by the spring expression language, e.g. 'public Optional<Adress> findBytownAndStreet(String town, String street);'
 - are the injected services hidden behind an interface (or is there a fragment of code, where the implementation of the interface is injected)
 - and several other complex checks, but these should be created by our project members.

### Our Questions

1. Where are the arguments of a yml file stored?
In the gatorgrader.yml file of java-assigment-starter are a few java-file-related arguments, e.g. "--exists" below SampleLabMain.java.
I also found the matching python file (check_ConfirmFileExists.py), but I can't find any connection between them. How does gatorgrader know what to do with this instruction?
I'm asking because we probably will have to add custom arguments and it would be helpful to understand how the existing arguments work.

2. Is it possible to include unit tests in the grading task, or does that have to be a separate gradle task?
		
3. How can we configure the gatorgradle task for multiple projects?
In this scenario the deadline for a task is over and every project shall be graded.
Assuming all projects have been checked out into a local directory, (how) can we use a single gatorgrader.yml file to grade all these projects?
What we try to achieve is:

- that the projects should not include the yml file, as the student might adjust his project in a way, that the gradle task scores 100%, while being not functional at all.
- the gradle task is called once, with a single yml file and a parameter, on which directory it should run (the one with all projects).

4. Can the reports be stored in a textfile?
Since we are not familiar with the functionality of Linux, this might be a off topic question, but is it possible to store the results, which is currently only displayed in a terminal window, in a textfile? This could be used to generate a report for all projects of a task, it could also be used later to send the results of the grading to a student via email.

That's it. I apologize for the wall of text, but I assume it is better to have too many informations than too few. If you need any other information, please let me know
 
Thank you for your help and kind regards from germany

André

**Comment by Michionlion** [comment]
I can answer some of these questions, however others may need to weigh in as well.

1) The pipeline from `gatorgrader.yml` to python code being run is as follows:

* `gatorgrader.yml` parsed in `org.gatorgradle.config.GatorGradleConfig`
* Individual python grading tasks (lines from `gatorgrader.yml`) are executed in `org.gatorgradle.task.GatorGradleTask` (each represented by a `Command`)
* GatorGrader loads the available plugins in `gator/orchestrate.py#L93`
* GatorGrader finds the requested check in `gator/orchestrate.py#L104`
* GatorGrader executes the check in `gator/orchestrate.py#L120`, calling the `act` method of the check's module
* The check's report is passed back up the call chain, eventually landing in `org.gatorgradle.display.CommandOutputSummary`, and printed.

Keep in mind that this process recently changed a lot in `v1.0.0`, although the basic mechanisms are the same (just no "plugin" concept, and instead hard-coded checks). The `check_ConfirmFileExists.py` module you mention is just such a plugin, and is loaded by `gator/orchestrate.py` from it's source code, if the check's name is used in the `gatorgrader.yml` file.

2) A grading task such as `ExecuteCommand --command "gradle test"` (if Gradle was used as the build tool for the project -- any console command could be entered instead, such as `make test`, etc.) could ensure that all unit tests pass, and `MatchCommandRegex --command "gradle --console plain test" --count 6 --regex "Test \d+"` could check for a certain minimum amount of unit tests (or even more complex logic with more advanced regular expressions, like certain testing output or desired logging).

3) There is no current mechanism by which to do this, although it is an interesting use case. It would likely be easiest to simply put together a hack-ish script that would copy the `gatorgrader.yml` file, as well as any other configuration and `build.gradle` additions needed, to each project, and then run `gradle grade` in that project.

However, this is a slightly different use case than what GatorGrader is designed for: it is meant to act as a student-guide, where completing the checks guides a student through an assignment, and showing students the checks in no way makes the assignment easier. There are many ways that your teacher could structure assignments such that security through obscurity is not needed, such as requiring certain functionality tested by an `expect`-based testing script, or simply creating a unit testing suite themselves that the students must make pass. Often, when students do not understand the requirements they are meant to complete (especially if the requirements are complex, such as certain types of expressions whose effects are possible to generate in a multitude of ways), they are very confused by an eventual grade from GatorGrader, despite having fully completed the assignment to a human teacher's satisfaction, because GatorGrader must account for all possible solutions. This can be fairly easy or very difficult depending on the particular assignment and how it could be tested.

A simple answer: This is an interesting use case, and we are interested in supporting it, but it does not currently exist. As a note, users currently design assignments so that it is not possible to score 100% while being non-functional. We would be very interested in learning more about this use case and reasoning behind it!

4) Executing GatorGradle like `gradle grade --console=plain --quiet > results.txt` would place the console output (with most gradle output removed, given `--quiet`) in the `results.txt` file. This uses file redirection, an easy way to manipulate standard output. It should be possible on all operating systems with various different syntax. An aside: it may be useful to use `gradle grade |& tee results.txt` on Linux operating systems -- piping into `tee` splits the output, writing to both the console and file and allowing the user to watch output in real time while storing it. More information on [redirection](https://www.guru99.com/linux-redirection.html) and [piping](https://linuxhint.com/linux_pipe_command/).

Of course, these answers are not exhaustive, and I look forward to having a dialogue as your team explores GatorGrader and it's ecosystem more! Be sure to check out the [example repositories](https://github.com/GatorEducator/java-assignment-solution-100-01), which will probably help with understanding the basics. Don't be afraid to reach out with more questions as well!

**Comment by Andre-Forberg** [comment]
Hi Saejin,

thank you for your detailed answer, this will help us a lot!
I will try to answer each of your questions one by one:

1. Since we are only planning to create a working prototype, the version 0.2.0 will work just fine for our purposes. Updating to version 1.0.0 will probably not be on our schedule, since we only have 8 weeks left (excluding christmas time and end of year). The project will end at the end of january (and we are still in the 'analysis & evaluation' phase).

2. If I understood you correctly, there is no built-in report functionality (like the one for java classes) for unit tests, and there also is no command for calling unit tests in the yml file, right?
Your approach would be to create a separate gradle task, which executes the unit tests. 
We've got an example project and we will try to create some typical checks for this. This might take a while, since we only know the gradle basics, but this is our part, we will make this work.

3. The approach to copy the yml file into the projects seems to be the best solution in this scenario. 
The idea with this one was to keep the project 'clean', which means there should be no extra files other than the project relevant ones. There are already 5 different gradle based files and an extra gradle folder which contains a wrapper jar, 3 different source folders (one for java classes, one for tests and one for resources like html files), so on top of that there would be another folder 'config' which has nothing to do with the project configuration itself. I know that 'config' is just a folder name, which can be changed, but I hope you get my point.
Anyway, since we are only working on a prototype, this will do the trick.

The use case that a student might read the yml file and adjust his project was only an idea from me (doesn't come from the professor), and my thought process was, that either the student deletes some checks, or:
 - the student looks at the yml file 
 - he sees that there is a check for logging mechanisms,
 - he will then write some logging calls in methods so that one extra test will pass
 - he got now 5/9 tests passing, which is the minimum required amount of tests to pass this project, while haven't done anything particularly to solve the problem.
On the other hand, since the student knows what the teacher will grade, he will only work on that parts, while loosing track of the actual goal of the project or not coming up with own ideas of improvement.
(This was a hard one to phrase, if you are not a native english speaker).

TL;DR: For our prototype:
 - We will include the gatorgrader files in a project 
 - the student can see what the teacher is grading
 - the file will be readonly, so students can't delete some checks 
 - a script will call the grade task for each project.
 
4. We didn't had time to try this or look at your suggested readings, but we will do this for the rest of the week.

Our weekly meeting is on friday, eventually I will write a follow up, if gatorgrader will be used for the prototype or if another tool will be the desired one. 
Thanks for your time and have a great week!

André

**Comment by Michionlion** [comment]
Per 2), any existing unit testing framework can be used -- including just a bare Java class with a main method, so whatever you use for unit tests currently can be substituted for the `gradle test` command I put as an example. We do not explicitly parse unit test results, but check whether a suite passes based on the exit code of whatever command is used to run it.

Per 3), while the Gradle wrapper is sometimes useful, we've found that in general it may be best to have Gradle installed on student's computers, which will reduce the complexity of all those scripts and folders in the project. This, however, is a personal choice. All of this is heavily dependent on your preferences, so I totally understand.

**Comment by schultzh** [comment]
Hello! We are delighted by your interest in the GatorGrader tool! We ask that you keep all discussions and questions in the issue tracker so those currently tasked with developing the tool can also be included in the discussion and contribute. We are also students and do our best to respond in a timely manor within 24 to 48 hours. Thank you!

**Comment by corlettim** [comment]
To follow up with @schultzh message, if anything has been discussed over email, we ask that you please place those comments here! Thank you!

**Comment by gkapfham** [comment]
Hello @Andre-Forberg is there anything else that you would like to discuss in this issue? Can we close this issue now? Thanks for your feedback, it is greatly appreciated.

**Comment by Andre-Forberg** [comment]
Hi everyone,

I apologize for not keeping you up to date. 
Last Friday we finished the project, but we didn't reached the prototyping phase. So we summarized our results into one report for our professors. Gatorgrader might be used in future projects, so someone else might contact you again later this year for further questions. 
Our proposed solution is the integration of gatorgrader in an web based grading application, as an addition to the basic functionality of the grading application.

Thank you all for your time, your very detailed replies to my questions, it helped us a lot.
Currently we don't have anymore questions, so this issue can be closed.
